### PR TITLE
[DEE-456] nlp dummy model refactoring

### PR DESCRIPTION
**Pull request by yromanyshyn**
@Nadav-Barak 

> Currently we have a dummy model for tabular and another one for nlp but they basically do the same stuff.
We would like them to have common abstract containing most of the logic with child class implement datadype specific functionalities.

they do not have much in common, in general, `_predict` and `_predict_proba` methods could be separated into some kind of base class, but in this case, we would need to unify `nlp.TextData` and `tabular.Dataset` APIs, everything else is just a validation logic that is pretty specific IMO. So, I do not see a need in creating some kind of ambiguous and not clear abstractions just to reduce a few lines of code

In general, I just refactored the nlp dummy model and moved the validation logic into the `input_validation.py` module

### CODE REVIEW

File: deepchecks/nlp/context.py [hunk 1]
```diff
@@ -39,13 +39,14 @@
     'TTokenPred'
 ]
 
-from deepchecks.utils.validation import is_sequence_not_str
 
 TClassPred = t.Union[t.Sequence[t.Union[str, int]], t.Sequence[t.Sequence[t.Union[str, int]]]]
 TClassProba = t.Sequence[t.Sequence[float]]
+# TODO: is it correct, why tuple?
 TTokenPred = t.Sequence[t.Sequence[t.Tuple[str, int, int, float]]]
```
Comment (noamzbr) [inline deepchecks/nlp/context.py:46]: It's old and not correct any more, should be 
```suggestion
TTokenPred = t.Sequence[t.Sequence[t.Union[str, int]]]
```

File: deepchecks/nlp/context.py [hunk 2]
```diff
@@ -72,68 +73,85 @@ class _DummyModel(BasicModel):
...
-                 y_pred_test: TTextPred,
-                 y_proba_test: TTextProba,
-                 train: t.Union[TextData, None] = None,
-                 y_pred_train: TTextPred = None,
-                 y_proba_train: TTextProba = None,
-                 model_classes: list = None,
-                 validate_data_on_predict: bool = True):
+    def __init__(
+        self,
+        *,
```
Comment (noamzbr) [inline deepchecks/nlp/context.py:78]: why enable any args?

File: deepchecks/nlp/context.py [hunk 3]
```diff
@@ -72,68 +73,85 @@ class _DummyModel(BasicModel):
...
+                    dataset=dataset,
+                    predictions=y_pred,
+                    probabilities=y_proba,
+                    n_of_classes=len(model_classes)
+                )
+                if y_pred is not None:
+                    y_pred = np.array(y_pred, dtype='float')
+                elif y_proba is not None:
+                    y_pred = (np.array(y_proba) > multilabel_proba_threshold)
+                    y_pred = [np.array(model_classes)[pred] for pred in y_pred]
```
Comment (noamzbr) [inline deepchecks/nlp/context.py:124]: No that line should be deleted for multilabel, the y_pred should be a binary matrix

File: deepchecks/nlp/context.py [hunk 4]
```diff
@@ -72,68 +73,85 @@ class _DummyModel(BasicModel):
...
+                _validate_text_classification(
+                    dataset=dataset,
+                    predictions=y_pred,
+                    probabilities=y_proba,
+                    n_of_classes=len(model_classes)
+                )
                 if y_pred is not None:
-                    y_pred_dict = dict(zip(dataset.get_original_text_indexes(), y_pred))
-                    predictions.update({dataset.name: y_pred_dict})
+                    y_pred = np.array(y_pred, dtype='str')
```
Comment (noamzbr) [inline deepchecks/nlp/context.py:134]: I'm actually unsure why we do the casting to string. Is the label also cast to str?

File: deepchecks/nlp/input_validations.py [hunk 5]
```diff
@@ -231,3 +235,197 @@ def compare_dataframes(
         difference = None
 
     return DataframesComparison(common, difference)
+
+
+def _validate_text_classification(
+    *,
```
Comment (noamzbr) [inline deepchecks/nlp/input_validations.py:241]: again why we need arbitrary arg?

File: deepchecks/nlp/input_validations.py [hunk 6]
```diff
@@ -231,3 +235,197 @@ def compare_dataframes(
         difference = None
 
     return DataframesComparison(common, difference)
+
+
+def _validate_text_classification(
```
Comment (noamzbr) [inline deepchecks/nlp/input_validations.py:240]: I would imagine that the difference between these and tabular _validate_classification would be that tabular has also binary. Is there a way to unify this multiclass logic with the tabular multiclass logic?

File: deepchecks/nlp/context.py (hunk 2 above)
Comment (yromanyshyn) [reply deepchecks/nlp/context.py:78]: sorry, not sure whether I get your question correctly, 
`*` is used to define keyword-only parameters ([docs](https://peps.python.org/pep-0570/#keyword-only-arguments)),
parameters that can be passed only as keywords 

IMO it will increase readibility

File: deepchecks/nlp/context.py (hunk 4 above)
Comment (yromanyshyn) [reply deepchecks/nlp/context.py:134]: > I'm actually unsure why we do the casting to string.

I also do not know/

> Is the label also cast to str?

yeap

```
# input_validations.py::validate_modify_label func, line 72
labels = [None if pd.isna(x) else str(x) for x in labels]
```

File: deepchecks/nlp/input_validations.py (hunk 5 above)
Comment (yromanyshyn) [reply deepchecks/nlp/input_validations.py:241]: > sorry, not sure whether I get your question correctly,
> `*` is used to define keyword-only parameters ([docs](https://peps.python.org/pep-0570/#keyword-only-arguments)),
> parameters that can be passed only as keywords
> 
> IMO it will increase readibility

File: deepchecks/nlp/context.py (hunk 1 above)
Comment (yromanyshyn) [reply deepchecks/nlp/context.py]: fixed

File: deepchecks/nlp/context.py [hunk 7]
```diff
@@ -39,13 +39,14 @@
...
-from deepchecks.utils.validation import is_sequence_not_str
 
 TClassPred = t.Union[t.Sequence[t.Union[str, int]], t.Sequence[t.Sequence[t.Union[str, int]]]]
 TClassProba = t.Sequence[t.Sequence[float]]
+# TODO: is it correct, why tuple?
 TTokenPred = t.Sequence[t.Sequence[t.Tuple[str, int, int, float]]]
 TTextPred = t.Union[TClassPred, TTokenPred]
-TTextProba = t.Union[TClassProba]  # TODO: incorrect, why union have only one type argument?
+# TODO: incorrect, why union have only one type argument?
+TTextProba = t.Union[TClassProba]
```
[the remark this answers is not in the document]
Comment (yromanyshyn) [reply deepchecks/nlp/context.py]: fixed

File: deepchecks/nlp/context.py (hunk 3 above)
Comment (yromanyshyn) [reply deepchecks/nlp/context.py]: I replaced it with 
```
y_pred = (np.array(y_proba) > multilabel_proba_threshold)
y_pred = y_pred.astype(int)
```

File: deepchecks/nlp/input_validations.py (hunk 6 above)
Comment (yromanyshyn) [reply deepchecks/nlp/input_validations.py:240]: > Is there a way to unify this multiclass logic with the tabular multiclass logic?

and Yes and No/

you know, IMO, it is one of those cases when code indeed has some similarities but differ in details and it is  not obvious how to separate those similarities into generic enough function, and if you will try you will end up with a code that is less readable and is harder to understand

for example, I do not find error messages in the `_validate_text_classification` to be suited for the tabular case, so should a unified function accept error messages templates?/

IMO, not worth overcomplicating very simple code just to reduce a few lines

File: deepchecks/nlp/context.py (hunk 4 above)
Comment (noamzbr) [reply deepchecks/nlp/context.py]: 🤷‍♂️

**Review by noamzbr — APPROVED** [review]

File: deepchecks/nlp/input_validations.py [hunk 8]
```diff
@@ -236,3 +240,209 @@ def compare_dataframes(
...
```
Comment (Nadav-Barak) [inline deepchecks/nlp/input_validations.py:245]: _validate_text_classification_predictions
so its clear it does not validate anything else

File: deepchecks/nlp/input_validations.py [hunk 9]
```diff
@@ -236,3 +240,209 @@ def compare_dataframes(
...
+        )
+        if not is_sequence_not_str(predictions):
+            raise ValidationError(format_error_message)
+        if len(predictions) != dataset.n_samples:
+            raise ValidationError(
+                f'Check requires predictions for the "{dataset.name}" dataset '
+                f'to have {dataset.n_samples} rows, same as dataset'
+            )
+        try:
+            predictions = np.array(predictions, dtype='str')
```
Comment (Nadav-Barak) [inline deepchecks/nlp/input_validations.py:269]: will not work well with None in predictions, better to use 'object'

File: deepchecks/nlp/input_validations.py (hunk 8 above)
Comment (yromanyshyn) [reply deepchecks/nlp/input_validations.py:245]: it validates not only predictions

**Comment by yromanyshyn** [comment]
@Nadav-Barak 

> To verify, either way predictions and probabilities return as np array?

yeap

File: deepchecks/nlp/input_validations.py (hunk 9 above)
Comment (Nadav-Barak) [reply deepchecks/nlp/input_validations.py:269]: if you can add test with None in the prediction for regular classification to verify everything works

Comment (yromanyshyn) [reply deepchecks/nlp/input_validations.py]: added
